.github/workflows/build.yml

```
@@ -148,135 +165,233 @@ jobs:
           fi
           eval "$shell"
 
-      # TODO signing prototype (see above where "shouldSign" is populated)
+      - name: Generate Checksum
+        id: checksum
+        run: |
+          cd build
+          tar -cvf temp.tar -C temp .
+          echo "sha256=$(sha256sum temp.tar)" >> "$GITHUB_OUTPUT"
+      - name: Stage artifact
+        id: oci
+        uses: actions/upload-artifact@v4
+        with:
+          name: build-oci
+          path: |
+            build/temp.tar*
+          retention-days: 5
+
+  sign:
+    name: Sign
+    needs: build
+    if: fromJSON(needs.build.outputs.buildJson).shouldSign
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      actions: write # for https://github.com/andymckay/cancel-action (see usage below)
+      id-token: write # for AWS KMS signing (see usage below)
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          sparse-checkout-cone-mode: 'false'
+          sparse-checkout: |
+            .scripts/oci.jq
+            .scripts/provenance.jq
+      - name: Download a single artifact
+        uses: actions/download-artifact@v4
+        with:
+          name: build-oci
+      - name: Verify artifact
+        run: |
+          echo "${{ needs.build.outputs.sha256 }}" sha256sum -c
+          mkdir -p temp
+          tar -xvf temp.tar -C temp
```

- `${{ ... }}` inside a `run:` block is fundamentally unsafe and should always pass via `env:`
- now `temp.tar` includes our directory (and `provenance.json`), so no need to `mkdir`
- validate we got the files we expected back out

```suggestion
      - name: Extract Artifact
        env:
          sha256: ${{ needs.build.outputs.artifactSha256 }}
        run: |
          sha256sum <<<"$sha256 *temp.tar" --strict --check -
          tar -xvf temp.tar
          [ -d temp ]
          [ -s provenance.json ]
          jq 'empty' provenance.json # basic "valid JSON" check
```